### PR TITLE
 Introduce `config/application/` directory

### PR DESCRIPTION
**Pull request by casperisfine**
Ref: https://github.com/rails/rails/pull/44980

Historically, configs were supposed to be entirely set from `application.rb` and the various `config/environments/` files.

But Rails itself broke this assumption by setting some configurations from initializer files. Actually all the default initializer files are setting configurations.

The problem with this is that it becomes very hard to know when a specific configuration is considered final and is applied. This has led to many configurations being specifically applied on `after_initialize` so that they can be defined in initializers (See https://github.com/rails/rails/pull/44980 as one of the many example).

This is bad for several reasons:

  - We need an ad hoc fix every time.
  - It's totally unclear for users when their config actually takes effect.
  - It cause some other initializers to need to be delayed as well.
  - If you get it wrong, your config might silently not be applied.

The idea here is two fold.

First we introduce `config/application` so that we better support the pattern of breaking down `application.rb` in smaller initialization files, which seems to be the original need here.

Then in a followup PR, we can warn or error out when the `config` object is mutated past the point where it no longer is applied. This can be implemented as a new framework default for 7.1, e.g. (`config.freeze = true`).

### CODE REVIEW

**Comment by jonathanhefner** [comment]
This might be a bad idea, but what about supporting multiple files in `config/environments`?

For example, before running initializers, load config from `config/environments/#{environment}.rb` + `config/environments/all/*.rb` + `config/environments/#{environment}/*.rb` for a given `environment`.  (`new_framework_defaults.rb` would likely go into `config/environments/all/`.)

It would not alleviate the need to write e.g. `after_initialize`, but I feel like it might make the directory structure a bit more obvious.

**Comment by ghiculescu** [comment]
Yeah, my suggestion was `config/preinitializers` - similar rationale to ^. I don't love the idea of magic folders inside `config/initializers`.

**Comment by p8** [comment]
> Yeah, my suggestion was config/preinitializers - similar rationale to ^.

We used to have a config/preinitializer.rb that was run before any of the frameworks were loadedn 🙂 
https://github.com/rails/rails/commit/bd69589143371254b6d71aa74b1230144eb30a9c

**Comment by fxn** [comment]
Unfortunately I don't have much bandwidth to be able to think about this properly.

Intuitively, I like @jonathanhefner's proposal. I had a similar idea in mind.

Point is, the main driver of this refactor is to move configuration out of initializers and make effective the contract "you configure here. Later, everyone can assume it is all set". So, there's two concepts in my view:

* Configuration: Configure the application.
* Initializers: Run arbitrary code when the app boots.

It is the case that configuration runs (should run) before initializers run. But that reduces this to something weak, equates configuration to a hook. In my opinion, this is more specific than that: Configuration is not a mere initialization hook, it is a step of its own whose order of execution is managed by the framework.

So, `config/environments/all.rb` would be a substitute of `config/application.rb` as far as configuration is concerned. If we need a directory... so be it. But if not, the user can always split things to their liking.

Providing a better interface for "execute on boot an on each reload" may have an ad-hoc solution, and perhaps we do not need a generic directory-based interface to hooks. (I don't know.)

That is my intuition, just that.

**Comment by casperisfine** [comment]
I'm totally open to discussions on the directory structure, it's mostly proof of concept at this stage.

If we're going with a way to split `application.rb / environments/`, then my first thought would be:

  - load `config/application.rb`
  - load `config/application/**/*.rb`
  - load `config/environments/$RAILS_ENV.rb`
  - load `config/environments/$RAILS_ENV/**/*.rb`

As for allowing to skip  the `after_inititialize / to_prepare` and such, we can delay this until after we'd made this change to see how useful it seems.

Is everyone on board with the above?

One little thing I'm not certain about it wether the files in these extra directories should have an implicit `Rails.application.configure {}` around them or not.

**Comment by fxn** [comment]
@casperisfine I like that too. I believe we are moving in the right direction.

Good point re `Rails.application.configure`. If we make it implicit, we are telling the user clearly: the intention of these files is not to run anything on boot, use require_relative for that. This is specifically a place to configure your application. At first sight, that sounds good to me.

At the same time, there's an asymmetry with the parent files, ..., but being intentional is kind of more important to me compared with the current state of things.

**Comment by casperisfine** [comment]
Ok, I decided to go with baby steps so that's it's easier to reason about.

For now I only introduce `config/application/**/*.rb`, which is the perfect place for all the existing default initializers.

If everyone is good with this we can merge and then see if `config/environments/$RAILS_ENV/**/*.rb` is actually worth it.

Also as a follow up I'd like to introduce config freezing, so that it's more obvious for users if they set a configuration too late.

**Comment by jonathanhefner** [comment]
> see if `config/environments/$RAILS_ENV/**/*.rb` is actually worth it.

My motivation for this (and what prompted https://github.com/rails/rails/pull/44996#issuecomment-1115369433) is that I would like to eventually revisit https://github.com/rails/rails/pull/39498#issuecomment-638464893.  Specifically, I would like to make framework installers such as `rails active_storage:install` be self-contained.  Currently, we [generate framework config](https://github.com/rails/rails/blob/e4ec2cbb494f6ec7dda227b05ba8a1ec8675c796/railties/lib/rails/generators/rails/app/templates/config/environments/development.rb.tt#L37-L41) only when a new app is created, and regardless of whether the user will run the corresponding installer.

Of course, we could make the installer modify the relevant `config/environments/$RAILS_ENV.rb` files, but I feel like it's cleaner to generate e.g. `config/environments/$RAILS_ENV/active_storage.rb`.

But I agree there's no need to do it in this PR.  And if you don't have time to pursue it, I can look into it.

**Comment by casperisfine** [comment]
@jonathanhefner Makes sense. At this stage I think it's useful. I just prefer to do this in step to make it easier to get buy-in.

**Comment by ghiculescu** [comment]
I think https://github.com/Shopify/rails/blob/refactor-initializers/railties/lib/rails/generators/rails/app/templates/config/application/inflections.rb.tt should probably stay in initializers since it doesn't call `Rails.application.config....`, what do you think?

All the other ones that are currently created in a new Rails app, I agree should live in `config/application`.

**Comment by casperisfine** [comment]
> should probably stay in initializers since it doesn't call `Rails.application.config....`, what do you think?

Yeah, I actually wondered the same thing. I moved it because I thought the inflection config might have an impact early during boot. But it has been working so far, so I think you are right. I'll move it back.

**Comment by fxn** [comment]
If we are strict, inflections are configuration in the following sense.

The `once` autoloader is setup after configuration has been processed. `autoload_once_paths` is frozen, and the application by contract is allowed to autoload in `config/initializers` constants managed by that loader. This also affects to railtes (see [this one](https://github.com/hotwired/turbo-rails/blob/b9b87fc627d9148899970deea01069875f15db3f/lib/turbo/engine.rb#L39) for example).

Since by default, both loaders inflect with AS, to have everything squared and consistently well-ordered, inflections would go earlier in my view.

File: railties/lib/rails/generators/rails/app/app_generator.rb [hunk 1]
```diff
@@ -116,12 +117,12 @@ def config
     def config_when_updating
```
Comment (matthewd) [inline railties/lib/rails/generators/rails/app/app_generator.rb:117]: Does this need to handle both the old and new paths?

Comment (casperisfine) [reply railties/lib/rails/generators/rails/app/app_generator.rb:117]: That's something I was wondering. If these files were edited to contain things others than config sets, they shouldn't be moved.

**Comment by dhh** [comment]
I can recognize the lack of clarity around when initializers are run, and what's the order. That does seem like a good problem to tackle. But I don't recognize the premise presented at the top:

> Historically, configs were supposed to be entirely set from application.rb and the various config/environments/ files.
> 
> But Rails itself broke this assumption by setting some configurations from initializer files. Actually all the default initializer files are setting configurations

The original reason for adding config/initializers was indeed explicitly to break down application.rb configuration. It's also for configuring other settings, and I can see how that some times means we have a timing issue, but I just want to be clear about the fact that there's no broken assumption from my perspective here.

I don't really understand config/application, to be honest. I'd rather attempt to deal with the timing issue by continuing to build on our existing structure: Initializers are files that configure things. Initialize them. Maybe we could have something like `config/initializers/before` and `config/initializers/after`. Splitting the directory structure into two dedicated buckets for timing reasons.

**Comment by casperisfine** [comment]
> I don't recognize the premise presented at the top
I took it from what @fxn said on Basecamp:

> The main idea in the boot refactor in Rails 3 was: You configure in application.rb/environments, then railties/engines initializers run and use the app configuration to configure themselves, then application config/initializers run later.
But this idea has never been executed 100% with clarity and there are things that can be configure later and still have effect
> Pursuing the goal that you can configure anytime does not quite square with that design and could give false expectations, I think. Would leak.
> To make it more complicated, at some point it was said that you are discouraged from putting things in config/application.rb and you should configure in app initializers, this was inconsistent with the design.

> I don't really understand config/application, to be honest.

The idea is to introduce a strong separation between when the various `config.foo =` are set (`config/application` and `config/environments`), and when they are applied `initializers`. And to potentially enforce it by making the `config` object immutable once we start running initializers, so that's it's no longer possible to set a config past the point where it is applied.

**Comment by dhh** [comment]
Gotcha. Probably a lot of different ideas of how this was intended. What I've had in mind all along with `config/initializers` was a way to initialize and configure the entire app – your code, config, gems – as logical concerns. And that this would be cross-env cutting.

Totally buy the need to control the order that these are run in a better way. But to me these are still all just initializers, and what we're trying to solve is a clear way to set when they run. Not keen to introduce another conceptual model in addition to that.

**Comment by casperisfine** [comment]
If we don't want to introduce new things, then a potential solution would be to run `config/initializers` much earlier, if not as the very first initializer.

<details>
 <summary>I did a quick hack to show the initializers orders and a new empty app, the `==========` line is the first `config/initializers` file:</summary>

```
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_config]
[:load_environment_hook]
[:load_active_support]
[:set_eager_load]
[:initialize_logger]
[:initialize_cache]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_load_path]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_autoload_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:set_eager_load_paths]
[:setup_once_autoloader]
[:bootstrap_hook]
[:set_secrets_root]
["active_support.isolation_level"]
["active_support.remove_deprecated_time_with_zone_name"]
["active_support.set_authenticated_message_encryption"]
["active_support.reset_execution_context"]
["active_support.reset_all_current_attributes_instances"]
["active_support.deprecation_behavior"]
["active_support.initialize_time_zone"]
["active_support.initialize_beginning_of_week"]
["active_support.require_master_key"]
["active_support.set_error_reporter"]
["active_support.set_configs"]
["active_support.set_hash_digest_class"]
["active_support.set_key_generator_hash_digest_class"]
["active_support.set_rfc4122_namespaced_uuids"]
["active_support.set_fallback_to_marshal_deserialization"]
["active_support.set_default_message_encryptor_serializer"]
["active_support.set_default_message_verifier_serializer"]
["active_support.set_marshal_serialization"]
["action_dispatch.configure"]
["active_model.secure_password"]
["active_model.i18n_customize_full_message"]
["action_controller.assets_config"]
["action_controller.set_helpers_path"]
["action_controller.parameters_config"]
["action_controller.set_configs"]
["action_controller.compile_config_methods"]
["action_controller.request_forgery_protection"]
["action_controller.query_log_tags"]
["action_controller.test_case"]
["active_record.initialize_timezone"]
["active_record.logger"]
["active_record.backtrace_cleaner"]
["active_record.migration_error"]
["active_record.database_selector"]
["active_record.shard_selector"]
["Check for cache versioning support"]
["active_record.check_schema_cache_dump"]
["active_record.define_attribute_methods"]
["active_record.warn_on_records_fetched_greater_than"]
["active_record.sqlite3_production_warning"]
["active_record.set_configs"]
["active_record.initialize_database"]
["active_record.log_runtime"]
["active_record.set_reloader_hooks"]
["active_record.set_executor_hooks"]
["active_record.add_watchable_files"]
["active_record.clear_active_connections"]
["active_record.set_filter_attributes"]
["active_record.set_signed_id_verifier_secret"]
["active_record_encryption.configuration"]
["active_record.query_log_tags_config"]
["active_record.unregister_current_scopes_on_unload"]
["global_id"]
["active_job.logger"]
["active_job.custom_serializers"]
["active_job.set_configs"]
["active_job.set_reloader_hook"]
["active_job.query_log_tags"]
["action_mailer.logger"]
["action_mailer.set_configs"]
["action_mailer.set_autoload_paths"]
["action_mailer.compile_config_methods"]
["test_unit.line_filtering"]
[:set_default_precompile]
[:quiet_assets]
[:asset_url_processor]
[:asset_sourcemap_url_processor]
[:jbuilder]
["web_console.initialize"]
["web_console.development_only"]
["web_console.insert_middleware"]
["web_console.mount_point"]
["web_console.template_paths"]
["web_console.permissions"]
["web_console.whiny_requests"]
["i18n.load_path"]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_routing_paths]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_locales]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:add_view_paths]
[:prepend_helpers_path]
[:prepend_helpers_path]
[:prepend_helpers_path]
[:prepend_helpers_path]
[:prepend_helpers_path]
[:prepend_helpers_path]
[:prepend_helpers_path]
["turbo.no_action_cable"]
["turbo.assets"]
["turbo.helpers"]
[:prepend_helpers_path]
[:prepend_helpers_path]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
[:load_config_initializers]
========================================
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:wrap_executor_around_load_seed]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:engines_blank_point]
[:append_assets_path]
["action_view.logger"]
["action_view.caching"]
["action_view.setup_action_pack"]
["action_view.collection_caching"]
[:append_assets_path]
["active_storage.configs"]
["active_storage.attached"]
["active_storage.verifier"]
["active_storage.services"]
["active_storage.queues"]
["active_storage.reflection"]
["action_view.configuration"]
["active_storage.asset"]
["active_storage.fixture_set"]
[:append_assets_path]
["action_cable.helpers"]
["action_cable.logger"]
["action_cable.asset"]
["action_cable.set_configs"]
["action_cable.routes"]
["action_cable.set_work_hooks"]
[:append_assets_path]
["action_mailbox.config"]
[:append_assets_path]
["action_text.attribute"]
["action_text.asset"]
["action_text.attachable"]
["action_text.helper"]
["action_text.renderer"]
["action_text.system_test_helper"]
["action_text.configure"]
[:append_assets_path]
["importmap"]
["importmap.reloader"]
["importmap.cache_sweeper"]
["importmap.assets"]
["importmap.helpers"]
["importmap.rescuable_asset_errors"]
[:append_assets_path]
["turbo.broadcastable"]
["turbo.mimetype"]
["turbo.renderer"]
["turbo.signed_stream_verifier_key"]
["turbo.test_assertions"]
["turbo.integration_test_request_encoding"]
[:append_assets_path]
["stimulus.assets"]
[:append_assets_path]
[:add_generator_templates]
[:setup_main_autoloader]
[:setup_default_session_store]
[:build_middleware_stack]
[:define_main_app_helper]
[:add_to_prepare_blocks]
[:run_prepare_callbacks]
[:eager_load!]
[:finisher_hook]
[:configure_executor_for_concurrency]
[:add_internal_routes]
[:set_routes_reloader_hook]
[:set_clear_dependencies_hook]
```
</details>

I'm not sure how deliberate it's current position here, but we could try to move it to way higher, maybe just after `:bootstrap_hook` ?

```ruby
      initializer :bootstrap_hook, group: :all do |app|
        ActiveSupport.run_load_hooks(:before_initialize, app)
      end
```

But I assume plenty of existing initializers would break if we did that.

**Comment by ghiculescu** [comment]
> But I assume plenty of existing initializers would break if we did that.

Thinking out loud, I guess existing initializers could break with this change if they depend on stuff that happens above those lines (and that doesn't use `after_initialize`). Examples could be things like locales, routes, or secrets being loaded. So it seems pretty likely this'll break something.

> But to me these are still all just initializers, and what we're trying to solve is a clear way to set when they run.

To me the issue is that there's some initializers you want to run *before* Rails initializes (those that set Rails configuration), and some that you want to run *after* Rails initializers (those that configure gems or your own code, and depend on Rails being fully initialized).

**Comment by dhh** [comment]
Don't think we can change the order of where the current initializers run. That's going to break things for existing apps.

An alternative to special folders would be a special top-level instruction inside the initializer, perhaps. Like `run_before_config_initialization!` and `run_after_config_initialization!`.

**Comment by byroot** [comment]
> a special top-level instruction inside the initializer,

I'd rather introduce another directory as it would be more obvious for a user which file are loaded in which order just looking at the file tree rather than having to open all the files one by one.

**Comment by rafaelfranca** [comment]
What prevents us to just free the `config` objects and not allow anything to happen after the initialization before introducing an escape hatch? The escape hatch is only necessary if application or library code early load one of the Rails components. In my opinion that is a bug in their code that should be fixed. I don't think we should be blessing this mistake by introducing an escape hatch.

**Comment by ghiculescu** [comment]
> The escape hatch is only necessary if application or library code early load one of the Rails components.

Hmm, I thought the issue was also if Rails configuration happens in initializers. It seems like it's been blessed plenty in the past (incidentally I just came across https://github.com/rails/rails/commit/614fc3d827579cebb26124bb5b574c045255c832).

**Comment by rafaelfranca** [comment]
> Hmm, I thought the issue was also if Rails configuration happens in initializers. It seems like it's been blessed plenty in the past.

It is fine if configurations happens in initializers. The problem only happens when some code loads the Rails components before those initializers are run. For example loading `ActiveRecord::Base` in one initialize and in another initializer setting `config.active_record.something`. For this case, the loading of `ActiveRecord::Base` should be avoided by using a `on_load(:active_record)` hook. That is how all the Rails components do things.

A new Rails app should not have any problem on setting config on initializers. As soon gems that are loading Rails components incorrectly, or application code also loading Rails components incorrectly get added, is when the configs in initializers can start to fail.

**Pull request closed without merging**